controllers/machinehealthcheck_targets.go

```diff
@@ -0,0 +1,259 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// EventDetectedUnhealthy is emitted in case a node associated with a
+	// machine was detected unhealthy
+	EventDetectedUnhealthy string = "DetectedUnhealthy"
+)
+
+// healthCheckTarget contains the information required to perform a health check
+// on the node to determine if any remediation is required.
+type healthCheckTarget struct {
+	Machine *clusterv1.Machine
+	Node    *corev1.Node
+	MHC     *clusterv1.MachineHealthCheck
+}
+
+func (t *healthCheckTarget) string() string {
+	return fmt.Sprintf("%s/%s/%s/%s",
+		t.MHC.GetNamespace(),
+		t.MHC.GetName(),
+		t.Machine.GetName(),
+		t.nodeName(),
+	)
+}
+
+// Get the node name if the target has a node
+func (t *healthCheckTarget) nodeName() string {
+	if t.Node != nil {
+		return t.Node.GetName()
+	}
+	return ""
+}
+
+// Determine whether or not a given target needs remediation.
+// The node will be need rememdiation if any of the following are true:
+// - The Machine has failed for some reason
+// - The Machine did not get a node before `timeoutForMachineToHaveNode` elapses
+// - The Node has gone away
+// - Any condition on the node is matched for the given timeout
+// If the target doesn't currently need rememdiation, provide a duration after
+// which the target should next be checked.
+// The target should be requeued after this duration.
+func (t *healthCheckTarget) needsRemediation(logger logr.Logger, timeoutForMachineToHaveNode time.Duration) (bool, time.Duration) {
+	var nextCheckTimes []time.Duration
+	now := time.Now()
+
+	// machine has failed
+	if t.Machine.Status.FailureReason != nil {
+		logger.V(3).Info("Target is unhealthy", "reason", t.Machine.Status.FailureReason)
+		return true, time.Duration(0)
+	}
+
+	// the node has not been set yet
+	if t.Node == nil {
+		// status not updated yet
+		if t.Machine.Status.LastUpdated == nil {
+			return false, timeoutForMachineToHaveNode
+		}
+		if t.Machine.Status.LastUpdated.Add(timeoutForMachineToHaveNode).Before(now) {
+			logger.V(3).Info("Target is unhealthy: machine has no node", "duration", timeoutForMachineToHaveNode.String())
+			return true, time.Duration(0)
+		}
+		durationUnhealthy := now.Sub(t.Machine.Status.LastUpdated.Time)
+		nextCheck := timeoutForMachineToHaveNode - durationUnhealthy + time.Second
+		return false, nextCheck
+	}
+
+	// the node does not exist
+	if t.Node != nil && t.Node.UID == "" {
+		return true, time.Duration(0)
+	}
+
+	// check conditions
+	for _, c := range t.MHC.Spec.UnhealthyConditions {
+		now := time.Now()
```

Comment:
Do you need to redeclare each time inside the loop or can you use `now` from L76?

Reply:
Good point, should be fine to use the one from L76